### PR TITLE
SB-791  Reimplement `ArtifactDirectoryOperation` using `RepositoryPath` abstraction

### PR DESCRIPTION
**Pull request by sbespalov**
Rebased with upstream/master

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactDirectoryOperation.java [hunk 1]
```diff
@@ -1,28 +1,29 @@
...
-import org.carlspring.strongbox.storage.Storage;
-import org.carlspring.strongbox.storage.repository.Repository;
-
-import java.io.File;
-import java.nio.file.Path;
+import java.io.IOException;
 import java.util.LinkedHashMap;
 import java.util.List;
 
+import org.carlspring.strongbox.providers.io.RepositoryFileSystem;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactDirectoryOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationGenerateChecksumOperation.java [hunk 2]
```diff
@@ -28,119 +22,90 @@
...
-    public LayoutProviderRegistry getLayoutProviderRegistry()
-    {
-        return layoutProviderRegistry;
-    }
-
-    public void setLayoutProviderRegistry(LayoutProviderRegistry layoutProviderRegistry)
-    {
-        this.layoutProviderRegistry = layoutProviderRegistry;
+        RepositoryPath basePath = versionDirectories.get(0).getParent();
+        RepositoryLayoutFileSystemProvider provider = (RepositoryLayoutFileSystemProvider) basePath.getFileSystem()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationGenerateChecksumOperation.java]: :+1: It looks like we should add  `null` check here.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationReportOperation.java [hunk 3]
```diff
@@ -22,65 +20,66 @@
...
+    public ArtifactLocationReportOperation(RepositoryPath basePath)
     {
         setBasePath(basePath);
     }
 
-    public void execute(Path path)
+    public void execute(RepositoryPath path) throws IOException
     {
-        File f = path.toAbsolutePath().toFile();
+        //File f = path.toAbsolutePath().toFile();
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationReportOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationReportOperation.java [hunk 4]
```diff
@@ -22,65 +20,66 @@
...
+
+        //String[] list = f.list(new PomFilenameFilter());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationReportOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java [hunk 5]
```diff
@@ -284,7 +286,7 @@ public void undelete(RepositoryPath path)
         else
         {
             Files.walkFileTree(trashPath.getTarget(),
-                               new MoveDirVisitor(trashPath.getTarget(),
+                               new MoveDirectoryVisitor(trashPath.getTarget(),
                                                   path.getTarget(),
                                                   StandardCopyOption.REPLACE_EXISTING));
         }
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java]: Thanks for renaming the visitor! The name looks better now. Could you just re-indent the method's parameter's above as well? :)

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java [hunk 6]
```diff
@@ -507,15 +508,15 @@ private Path unwrap(Path path)
...
 
         private final Path fromPath;
         private final Path toPath;
         private final CopyOption copyOption;
 
-        public MoveDirVisitor(Path fromPath,
+        public MoveDirectoryVisitor(Path fromPath,
                               Path toPath,
                               CopyOption copyOption)
         {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryFileSystemProvider.java]: Cosmetic: Could your please re-indent the parameters?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryPath.java [hunk 7]
```diff
@@ -211,14 +213,8 @@ public String toString()
     @Override
     public boolean equals(Object obj)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryPath.java]: :+1:

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryPath.java [hunk 8]
```diff
@@ -211,14 +213,8 @@ public String toString()
...
     {
-        return target.equals(obj instanceof Path ? unwrap((Path) obj) : obj);
-    }
-
-    @Override
-    public int hashCode()
-    {
-        return target.hashCode();
+        return  getTarget().equals(obj instanceof RepositoryPath ? unwrap((Path) obj) : obj);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/RepositoryPath.java]: :+1:

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 9]
```diff
@@ -103,35 +88,6 @@ public Storage getStorage(String storageId)
         return configurationManager.getConfiguration().getStorage(storageId);
     }
 
-    @Override
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: I understand you have removed both the `getInputStream` and `getOutputStream` methods. I don't think I fully understand the reasoning behind this, or that I agree/approve.

I think that the layout provider (whether the abstract implementation, or the actual implementation) should provide these methods for convenience. They can be doing the `Files.newInputStream` there, or whatever the logic should be. Right now we're talking about file-system like `InputStream`-s. What will we do in the future, if, for example, we decide to have a database provider (or something along these lines) where `Files.newInputStream` will probably not be applicable? A non-Strongbox-core developer/user of our code should be able to figure out how to do things with the least amount of effort and I think removing these methods that hide the underlying implementation will lead to some confusion in the future, as people will be wondering how to get an input/output stream and will end up doing their own thing, no necessarily being `Files.get[Input|Output]Stream`.

What are your thoughts on this?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 10]
```diff
@@ -147,146 +103,6 @@ protected Repository getRepository(String storageId,
...
-                                                    return;
-                                                }
-
-                                                result.getHexDigests().put(a, checksum);
-                                            });
-        }
-        return result;
-    }
-
-    private String getChecksum(String storageId,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: So where has this gone?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java [hunk 11]
```diff
@@ -499,72 +307,7 @@ public boolean containsPath(Repository repository,
 
         return Files.exists(repositoryPath.resolve(path));
     }
-
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: And where has this gone?

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/RepositoryLayoutFileSystemProvider.java [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        else
+        {
+            checksum = is.getMessageDigestAsHexadecimalString(digestAlgorithm);
+        }
+        return checksum;
+    }
+
+
+
+    protected RepositoryPath getChecksumPath(RepositoryPath path,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/RepositoryLayoutFileSystemProvider.java]: Feel free to remove the block of empty lines above.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/RepositoryLayoutFileSystemProvider.java [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+                                               t);
+                              }
+                          });
+        return result;
+    }
+    
+    public void storeChecksum(RepositoryPath basePath,
+                              boolean forceRegeneration)
+        throws IOException,
+        NoSuchAlgorithmException
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/RepositoryLayoutFileSystemProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java [hunk 14]
```diff
@@ -50,8 +53,9 @@ public ArtifactInputStream getInputStream(String storageId,
     {
         Repository repository = getConfiguration().getStorage(storageId).getRepository(repositoryId);
 
-        return getLayoutProviderRegistry().getProvider(repository.getLayout())
-                                          .getInputStream(storageId, repositoryId, path);
+        LayoutProvider layoutPtovider = getLayoutProviderRegistry().getProvider(repository.getLayout());
+        RepositoryPath repositoryPath = layoutPtovider.resolve(repository).resolve(path);
+        return (ArtifactInputStream) Files.newInputStream(repositoryPath);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/HostedRepositoryProvider.java]: Cosmetic: (Here and in a number of places below and across the pull) Could you please leave an empty line before `return` statements? Thanks!

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/ProxyRepositoryProvider.java [hunk 15]
```diff
@@ -146,9 +150,9 @@ public ArtifactOutputStream getOutputStream(String storageId,
         LayoutProvider layoutProvider = layoutProviderRegistry.getProvider(repository.getLayout());
 
         ArtifactCoordinates coordinates = layoutProvider.getArtifactCoordinates(artifactPath);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/repository/ProxyRepositoryProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/storage/FileSystemStorageProvider.java [hunk 16]
```diff
@@ -1,8 +1,6 @@
 package org.carlspring.strongbox.providers.storage;
 
-import java.io.FileNotFoundException;
 import java.io.IOException;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/storage/FileSystemStorageProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/storage/FileSystemStorageProvider.java [hunk 17]
```diff
@@ -1,8 +1,6 @@
...
-import java.io.InputStream;
 import java.io.OutputStream;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/storage/FileSystemStorageProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/storage/FileSystemStorageProvider.java [hunk 18]
```diff
@@ -51,62 +49,7 @@ public void register()
     }
 
     @Override
-    public OutputStream getOutputStreamImplementation(Path artifactPath)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/providers/storage/FileSystemStorageProvider.java]: In line with my earlier comments in this pull regarding the removal of the `get[Input|Output]Stream`, I'd like to hear your thoughts on this part here as well.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/AbstractMavenArtifactLocatorOperation.java [hunk 19]
```diff
@@ -24,79 +22,83 @@
...
 
     public AbstractMavenArtifactLocatorOperation()
     {
     }
 
-    public void execute(Path path)
+    public void execute(RepositoryPath path) throws IOException
     {
-        File f = path.toAbsolutePath().toFile();
+//        File f = path.toAbsolutePath().toFile();
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/AbstractMavenArtifactLocatorOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/AbstractMavenArtifactLocatorOperation.java [hunk 20]
```diff
@@ -24,79 +22,83 @@
...
 
-        String[] list = f.list(new PomFilenameFilter());
-        List<String> filePaths = list != null ? Arrays.asList(list) : new ArrayList<>();
+        List<Path> filePathList = Files.list(path)
+                                       .filter(p -> p.getFileName().toString().endsWith(".pom"))
+                                       .sorted()
+                                       .collect(Collectors.toList());
+        
+//        String[] list = f.list(new PomFilenameFilter());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/AbstractMavenArtifactLocatorOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/AbstractMavenArtifactLocatorOperation.java [hunk 21]
```diff
@@ -24,79 +22,83 @@
...
+            for (RepositoryPath directory : versionDirectories)
+            {
+                // We're using System.out.println() here for clarity and due to the length of the lines
+                System.out.println(" " + directory.toAbsolutePath());
             }
         }
+
+        RepositoryPath artifactPath = parentPath.getRepositoryRelative();
+        
+        //String artifactPath = parentPath.substring(getRepository().getBasedir().length() + 1, parentPath.length());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/AbstractMavenArtifactLocatorOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/MavenIndexerManagementOperation.java [hunk 22]
```diff
@@ -36,76 +35,81 @@ public MavenIndexerManagementOperation(RepositoryIndexManager repositoryIndexMan
...
+            return;
+        }
+        for (RepositoryPath versionDirectoryAbs : versionDirectories)
+        {
+            // We're using System.out.println() here for clarity and due to the length of the lines
+            // System.out.println(" " + versionDirectory.getAbsolutePath());
+
+            RepositoryPath artifactVersionDirectoryRelative  = versionDirectoryAbs.getRepositoryRelative();
+            
+//            String artifactVersionDirectory = versionDirectory.getPath();
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/MavenIndexerManagementOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/MavenIndexerManagementOperation.java [hunk 23]
```diff
@@ -36,76 +35,81 @@ public MavenIndexerManagementOperation(RepositoryIndexManager repositoryIndexMan
...
+//            String artifactVersionDirectoryRelative = artifactVersionDirectory.substring(getStorage().getRepository(repositoryId)
+//                                                                                                     .getBasedir()
+//                                                                                                     .length()
+//                    + 1,
+//                                                                                         artifactVersionDirectory.length());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/MavenIndexerManagementOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/MavenIndexerManagementOperation.java [hunk 24]
```diff
@@ -36,76 +35,81 @@ public MavenIndexerManagementOperation(RepositoryIndexManager repositoryIndexMan
...
+                     String extension = fileName.substring(fileName.lastIndexOf('.') + 1,
+                                                           fileName.length());
+
+                     artifact.setFile(f.toFile());
+
+                     try
+                     {
+                         indexer.addArtifactToIndex(repositoryId, f.toFile(), artifact);
+                     }
+                     catch (IOException e)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/MavenIndexerManagementOperation.java]: :-1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java [hunk 25]
```diff
@@ -49,83 +46,92 @@ public RemoveTimestampedSnapshotOperation(MavenSnapshotManager mavenSnapshotMana
         this.mavenSnapshotManager = mavenSnapshotManager;
     }
 
-    public void execute(Path path)
+    public void execute(RepositoryPath path) throws IOException
     {
-        File f = path.toAbsolutePath()
-                     .toFile();
+//        File f = path.toAbsolutePath()
+//                     .toFile();
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java [hunk 26]
```diff
@@ -49,83 +46,92 @@ public RemoveTimestampedSnapshotOperation(MavenSnapshotManager mavenSnapshotMana
...
+//
+//        String[] list = f.list(new PomFilenameFilter());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java [hunk 27]
```diff
@@ -49,83 +46,92 @@ public RemoveTimestampedSnapshotOperation(MavenSnapshotManager mavenSnapshotMana
...
+        
+        List<Path> filePathList = Files.walk(path)
+                .filter(p -> !p.getFileName().startsWith(".pom"))
+                .sorted()
+                .collect(Collectors.toList());
+        
+//        List<String> filePaths = list != null ? Arrays.asList(list) : new ArrayList<>();
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java [hunk 28]
```diff
@@ -49,83 +46,92 @@ public RemoveTimestampedSnapshotOperation(MavenSnapshotManager mavenSnapshotMana
...
+
+        RepositoryPath parentPath = path.getParent()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java]: Cosmetic: Please, re-indent, or put on the same line.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java [hunk 29]
```diff
@@ -49,83 +46,92 @@ public RemoveTimestampedSnapshotOperation(MavenSnapshotManager mavenSnapshotMana
...
+            RepositoryPath artifactPath = path.getRepositoryRelative();
 
-            // The current directory is out of the tree
-            if (previousPath != null && !parentPath.startsWith(previousPath))
-            {
-                getVisitedRootPaths().remove(previousPath);
-                previousPath = parentPath;
-            }
+//            String artifactPath = parentPath.substring(getRepository().getBasedir()
+//                                                                      .length() + 1, parentPath.length());
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/locator/handlers/RemoveTimestampedSnapshotOperation.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/services/impl/ArtifactMetadataServiceImpl.java [hunk 30]
```diff
@@ -115,12 +112,18 @@ public void rebuildMetadata(String storageId, String repositoryId, String basePa
         Storage storage = getConfiguration().getStorage(storageId);
         Repository repository = storage.getRepository(repositoryId);
 
-        basePath = basePath == null ? "/" : basePath;
+        //basePath = basePath == null ? "" : basePath;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/services/impl/ArtifactMetadataServiceImpl.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/storage/checksum/MavenChecksumManager.java [hunk 31]
```diff
@@ -1,199 +0,0 @@
-package org.carlspring.strongbox.storage.checksum;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/storage/checksum/MavenChecksumManager.java]: Where is the code moved?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java [hunk 32]
```diff
@@ -17,14 +17,26 @@
 import java.io.FilenameFilter;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java [hunk 33]
```diff
@@ -17,14 +17,26 @@
...
 import java.io.OutputStream;
+import java.nio.file.spi.FileSystemProvider;
 import java.security.NoSuchAlgorithmException;
 import java.util.Base64;
-import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import javax.annotation.PostConstruct;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java [hunk 34]
```diff
@@ -17,14 +17,26 @@
...
+import javax.inject.Inject;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java [hunk 35]
```diff
@@ -17,14 +17,26 @@
...
+
 import org.apache.commons.codec.digest.MessageDigestAlgorithms;
+import org.carlspring.strongbox.artifact.coordinates.ArtifactCoordinates;
+import org.carlspring.strongbox.artifact.coordinates.NugetHierarchicalArtifactCoordinates;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-nuget-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/NugetHierarchicalLayoutProvider.java]: :+1:

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/storage/checksum/MavenChecksumManager.java (hunk 31 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/storage/checksum/MavenChecksumManager.java]: This class was not used, so I removed it.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationGenerateChecksumOperation.java (hunk 2 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationGenerateChecksumOperation.java]: I think that `basePath` can not be `null` here.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java (hunk 9 above)
Comment (sbespalov) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/AbstractLayoutProvider.java]: As discussed in skype `getInput/OutputStream` methods was reverted as `final` in `AbstractLayoutProvider`. And this is just for convenience.

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationGenerateChecksumOperation.java (hunk 2 above)
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/handlers/ArtifactLocationGenerateChecksumOperation.java]: Okay

**Pull request merged**